app/controllers/application_controller.rb

```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
 
   private
 
+  set_current_tenant_through_filter
+  before_action :set_the_current_tenant
+  after_action  :unset_the_current_tenant
+
+  def set_the_current_tenant
+    return unless Tenant.tenancy_enabled?
+
+    tenant = Tenant.find_by(:external_tenant => user_identity)
+    if tenant
+      set_current_tenant(tenant)
+    else
+      error_document = {
+          "errors" => [
+              {
+                  "status" => "401",
+                  "detail" => "Account number not present or not found"
+              }
+          ]
+      }
+      render :json => error_document, :status => :unauthorized
+    end
+  end
+
+  def unset_the_current_tenant
+    set_current_tenant(nil)
+  end
+
+  def user_identity
+    ManageIQ::API::Common::Request.with_request(request) do |current|
```

Comment:
This should probably be around the entire request with a before_action (though if that's breaking a lot of things we can do in a followup)